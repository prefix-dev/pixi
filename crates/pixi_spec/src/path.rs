use std::path::Path;

use rattler_conda_types::{package::ArchiveIdentifier, NamelessMatchSpec};
use typed_path::{Utf8NativePathBuf, Utf8TypedPathBuf};

use crate::SpecConversionError;

/// A specification of a package from a git repository.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct PathSpec {
    /// The path to the package
    pub path: Utf8TypedPathBuf,
}

impl PathSpec {
    /// Converts this instance into a [`NamelessMatchSpec`] if the path points
    /// to binary archive.
    pub fn try_into_nameless_match_spec(
        self,
        root_dir: &Path,
    ) -> Result<Option<NamelessMatchSpec>, SpecConversionError> {
        if !self.is_binary() {
            // Not a binary package
            return Ok(None);
        }

        // Convert the path to an absolute path based on the root_dir
        let path = if self.path.is_absolute() {
            self.path
        } else if let Ok(user_path) = self.path.strip_prefix("~/") {
            let home_dir = dirs::home_dir()
                .ok_or_else(|| SpecConversionError::InvalidPath(self.path.to_string()))?;
            let Some(home_dir_str) = home_dir.to_str() else {
                return Err(SpecConversionError::NotUtf8RootDir(home_dir));
            };
            Utf8TypedPathBuf::from(home_dir_str)
                .join(user_path)
                .normalize()
        } else {
            let Some(root_dir_str) = root_dir.to_str() else {
                return Err(SpecConversionError::NotUtf8RootDir(root_dir.to_path_buf()));
            };
            let native_root_dir = Utf8NativePathBuf::from(root_dir_str);
            if !native_root_dir.is_absolute() {
                return Err(SpecConversionError::NonAbsoluteRootDir(
                    root_dir.to_path_buf(),
                ));
            }

            native_root_dir.to_typed_path().join(self.path).normalize()
        };

        // Convert the absolute url to a file:// url
        let local_file_url =
            file_url::file_path_to_url(path.to_path()).expect("failed to convert path to file url");

        Ok(Some(NamelessMatchSpec {
            url: Some(local_file_url),
            ..NamelessMatchSpec::default()
        }))
    }

    /// Converts this instance into a [`PathSourceSpec`] if the path points to a
    /// source package. Otherwise, returns this instance unmodified.
    #[allow(clippy::result_large_err)]
    pub fn try_into_source_path(self) -> Result<PathSourceSpec, Self> {
        if self.is_binary() {
            Err(self)
        } else {
            Ok(PathSourceSpec { path: self.path })
        }
    }

    /// Returns true if this path points to a binary archive.
    pub fn is_binary(&self) -> bool {
        self.path
            .file_name()
            .and_then(ArchiveIdentifier::try_from_path)
            .is_some()
    }
}

/// Path to a source package. Different from [`PathSpec`] in that this type only
/// refers to source packages.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct PathSourceSpec {
    /// The path to the package. Either a directory or an archive.
    pub path: Utf8TypedPathBuf,
}

impl From<PathSourceSpec> for PathSpec {
    fn from(value: PathSourceSpec) -> Self {
        Self { path: value.path }
    }
}
