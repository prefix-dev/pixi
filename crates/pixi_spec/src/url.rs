use rattler_conda_types::{package::ArchiveIdentifier, NamelessMatchSpec};
use rattler_digest::{Md5Hash, Sha256Hash};
use serde_with::serde_as;
use url::Url;

/// A specification of a package from a URL. This is used to represent both
/// source and binary packages.
#[serde_as]
#[derive(Debug, Clone, Hash, Eq, PartialEq, ::serde::Serialize, ::serde::Deserialize)]
pub struct UrlSpec {
    /// The URL of the package
    pub url: Url,

    /// The md5 hash of the package
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Md5>>")]
    pub md5: Option<Md5Hash>,

    /// The sha256 hash of the package
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Sha256>>")]
    pub sha256: Option<Sha256Hash>,
}

impl UrlSpec {
    /// Converts this instance into a [`NamelessMatchSpec`] if the URL points to
    /// a binary package.
    #[allow(clippy::result_large_err)]
    pub fn try_into_nameless_match_spec(self) -> Result<NamelessMatchSpec, Self> {
        if self.is_binary() {
            Ok(NamelessMatchSpec {
                url: Some(self.url),
                md5: self.md5,
                sha256: self.sha256,
                ..NamelessMatchSpec::default()
            })
        } else {
            Err(self)
        }
    }

    /// Converts this instance into a [`UrlSourceSpec`] if the URL points to a
    /// source package. Otherwise, returns this instance unmodified.
    #[allow(clippy::result_large_err)]
    pub fn try_into_source_url(self) -> Result<UrlSourceSpec, Self> {
        if self.is_binary() {
            Err(self)
        } else {
            Ok(UrlSourceSpec {
                url: self.url,
                md5: self.md5,
                sha256: self.sha256,
            })
        }
    }

    /// Returns true if the URL points to a binary package.
    pub fn is_binary(&self) -> bool {
        ArchiveIdentifier::try_from_url(&self.url).is_some()
    }
}

/// A specification of a source archive from a URL.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct UrlSourceSpec {
    /// The URL of the package
    pub url: Url,

    /// The md5 hash of the archive
    pub md5: Option<Md5Hash>,

    /// The sha256 hash of the archive
    pub sha256: Option<Sha256Hash>,
}

impl From<UrlSourceSpec> for UrlSpec {
    fn from(value: UrlSourceSpec) -> Self {
        Self {
            url: value.url,
            md5: value.md5,
            sha256: value.sha256,
        }
    }
}
