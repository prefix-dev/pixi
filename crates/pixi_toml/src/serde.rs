//! Provides a bridge to deserialize types that implement `serde::Deserialize` but not `toml_span::Deserialize`.

use std::collections::BTreeMap;

use serde::Deserialize;
use toml_span::{DeserError, ErrorKind, Value, value::ValueInner};

use crate::DeserializeAs;

/// A wrapper type that enables deserializing any type that implements `serde::Deserialize`
/// from a TOML value.
///
/// This is useful when you have types from external crates that already implement
/// `serde::Deserialize` but not `toml_span::Deserialize`.
///
/// # Example
/// ```ignore
/// use pixi_toml::{Serde, TomlWith};
///
/// // Assuming ExternalType implements serde::Deserialize but not toml_span::Deserialize
/// let value = th.optional::<TomlWith<_, Serde<ExternalType>>>("field")
///     .map(TomlWith::into_inner);
/// ```
pub struct Serde<T>(std::marker::PhantomData<T>);

impl<'de, T> DeserializeAs<'de, T> for Serde<T>
where
    T: Deserialize<'de>,
{
    fn deserialize_as(value: &mut Value<'de>) -> Result<T, DeserError> {
        // Convert TOML value to serde value
        let value_as_json = convert_toml_to_serde(value)?;

        // Deserialize using T's serde::Deserialize implementation
        T::deserialize(serde_value::ValueDeserializer::new(value_as_json)).map_err(
            |e: serde_value::DeserializerError| {
                DeserError::from(toml_span::Error {
                    kind: ErrorKind::Custom(e.to_string().into()),
                    span: value.span,
                    line_info: None,
                })
            },
        )
    }
}

/// Convert a TOML value to a serde value
pub fn convert_toml_to_serde(value: &mut Value) -> Result<serde_value::Value, DeserError> {
    Ok(match value.take() {
        ValueInner::String(s) => serde_value::Value::String(s.to_string()),
        ValueInner::Integer(i) => serde_value::Value::I64(i),
        ValueInner::Float(f) => serde_value::Value::F64(f),
        ValueInner::Boolean(b) => serde_value::Value::Bool(b),
        ValueInner::Array(mut arr) => {
            let mut json_arr = Vec::new();
            for item in &mut arr {
                json_arr.push(convert_toml_to_serde(item)?);
            }
            serde_value::Value::Seq(json_arr)
        }
        ValueInner::Table(table) => {
            let mut map = BTreeMap::new();
            for (key, mut val) in table {
                map.insert(
                    serde_value::Value::String(key.to_string()),
                    convert_toml_to_serde(&mut val)?,
                );
            }
            serde_value::Value::Map(map)
        }
    })
}
