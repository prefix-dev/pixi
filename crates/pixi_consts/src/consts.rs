use console::Style;
use lazy_static::lazy_static;
use std::fmt::{Display, Formatter};
use url::Url;

pub const DEFAULT_ENVIRONMENT_NAME: &str = "default";
pub const DEFAULT_FEATURE_NAME: &str = DEFAULT_ENVIRONMENT_NAME;
pub const PYPROJECT_PIXI_PREFIX: &str = "tool.pixi";

pub const PROJECT_MANIFEST: &str = "pixi.toml";
pub const PYPROJECT_MANIFEST: &str = "pyproject.toml";
pub const PROJECT_LOCK_FILE: &str = "pixi.lock";
pub const CONFIG_FILE: &str = "config.toml";
pub const PIXI_DIR: &str = ".pixi";
pub const PIXI_VERSION: &str = "0.29.0";
pub const PREFIX_FILE_NAME: &str = "pixi_env_prefix";
pub const ENVIRONMENTS_DIR: &str = "envs";
pub const SOLVE_GROUP_ENVIRONMENTS_DIR: &str = "solve-group-envs";
pub const PYPI_DEPENDENCIES: &str = "pypi-dependencies";
pub const DEPENDENCIES: &str = "dependencies";
pub const TASK_CACHE_DIR: &str = "task-cache-v0";
pub const PIXI_UV_INSTALLER: &str = "uv-pixi";
pub const CONDA_PACKAGE_CACHE_DIR: &str = rattler_cache::PACKAGE_CACHE_DIR;
pub const CONDA_REPODATA_CACHE_DIR: &str = rattler_cache::REPODATA_CACHE_DIR;
pub const PYPI_CACHE_DIR: &str = "uv-cache";
pub const CONDA_PYPI_MAPPING_CACHE_DIR: &str = "conda-pypi-mapping";
pub const CACHED_ENVS_DIR: &str = "cached-envs-v0";

pub const CONDA_INSTALLER: &str = "conda";

pub const ONE_TIME_MESSAGES_DIR: &str = "one-time-messages";

/// The default channels to use for a new project.
pub const DEFAULT_CHANNELS: &[&str] = &["conda-forge"];

pub const ENVIRONMENT_FILE_NAME: &str = "pixi";

lazy_static! {
    pub static ref TASK_STYLE: Style = Style::new().blue();
    pub static ref PLATFORM_STYLE: Style = Style::new().yellow();
    pub static ref ENVIRONMENT_STYLE: Style = Style::new().magenta();
    pub static ref FEATURE_STYLE: Style = Style::new().cyan();
    pub static ref SOLVE_GROUP_STYLE: Style = Style::new().cyan();
    pub static ref DEFAULT_PYPI_INDEX_URL: Url = Url::parse("https://pypi.org/simple").unwrap();
}

pub struct CondaEmoji;

impl Display for CondaEmoji {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if console::Term::stderr().features().colors_supported() {
            write!(f, "{}", console::style("C").bold().green())
        } else {
            write!(f, "(conda)")
        }
    }
}

pub struct PypiEmoji;

impl Display for PypiEmoji {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        if console::Term::stderr().features().colors_supported() {
            write!(f, "{}", console::style("P").bold().blue())
        } else {
            write!(f, "(pypi)")
        }
    }
}
