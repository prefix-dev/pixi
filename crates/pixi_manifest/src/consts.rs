use url::Url;

pub const PYPI_DEPENDENCIES: &str = "pypi-dependencies";
pub const DEFAULT_ENVIRONMENT_NAME: &str = "default";
pub const PROJECT_MANIFEST: &str = "pixi.toml";
pub const PYPROJECT_MANIFEST: &str = "pyproject.toml";
pub const DEFAULT_FEATURE_NAME: &str = DEFAULT_ENVIRONMENT_NAME;
pub const PYPROJECT_PIXI_PREFIX: &str = "tool.pixi";

lazy_static::lazy_static! {
    pub static ref DEFAULT_PYPI_INDEX_URL: Url = Url::parse("https://pypi.org/simple").unwrap();
}
