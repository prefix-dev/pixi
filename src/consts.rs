use console::Style;
use lazy_static::lazy_static;
use url::Url;

pub const PROJECT_MANIFEST: &str = "pixi.toml";
pub const PYPROJECT_MANIFEST: &str = "pyproject.toml";
pub const PROJECT_LOCK_FILE: &str = "pixi.lock";
pub const PIXI_DIR: &str = ".pixi";
pub const CONFIG_FILE: &str = "config.toml";
pub const PREFIX_FILE_NAME: &str = "pixi_env_prefix";
pub const ENVIRONMENTS_DIR: &str = "envs";
pub const SOLVE_GROUP_ENVIRONMENTS_DIR: &str = "solve-group-envs";
pub const PYPI_DEPENDENCIES: &str = "pypi-dependencies";
pub const TASK_CACHE_DIR: &str = "task-cache-v0";

pub const DEFAULT_ENVIRONMENT_NAME: &str = "default";

/// The default channels to use for a new project.
pub const DEFAULT_CHANNELS: &[&str] = &["conda-forge"];
pub const DEFAULT_FEATURE_NAME: &str = DEFAULT_ENVIRONMENT_NAME;

lazy_static! {
    pub static ref TASK_STYLE: Style = Style::new().blue();
    pub static ref PLATFORM_STYLE: Style = Style::new().yellow();
    pub static ref SOLVE_GROUP_STYLE: Style = Style::new().cyan();
    pub static ref DEFAULT_PYPI_INDEX_URL: Url = Url::parse("https://pypi.org/simple").unwrap();
}
