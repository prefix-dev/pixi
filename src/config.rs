use clap::{ArgAction, Parser};
use miette::{Context, IntoDiagnostic};
use rattler_conda_types::{Channel, ChannelConfig, ParseChannelError};
use serde::Deserialize;
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use url::Url;

use crate::consts;

/// Determines the default author based on the default git author. Both the name and the email
/// address of the author are returned.
pub fn get_default_author() -> Option<(String, String)> {
    let rv = Command::new("git")
        .arg("config")
        .arg("--get-regexp")
        .arg("^user.(name|email)$")
        .stdout(Stdio::piped())
        .output()
        .ok()?;

    let mut name = None;
    let mut email = None;

    for line in std::str::from_utf8(&rv.stdout).ok()?.lines() {
        match line.split_once(' ') {
            Some(("user.email", value)) => {
                email = Some(value.to_string());
            }
            Some(("user.name", value)) => {
                name = Some(value.to_string());
            }
            _ => {}
        }
    }

    Some((name?, email.unwrap_or_else(|| "".into())))
}

/// Get pixi home directory, default to `$HOME/.pixi`
///
/// It may be overridden by the `PIXI_HOME` environment variable.
///
/// # Returns
///
/// The pixi home directory
pub fn home_path() -> Option<PathBuf> {
    if let Some(path) = std::env::var_os("PIXI_HOME") {
        Some(PathBuf::from(path))
    } else {
        dirs::home_dir().map(|path| path.join(consts::PIXI_DIR))
    }
}

/// Returns the default cache directory.
/// Most important is the `PIXI_CACHE_DIR` environment variable.
/// - If that is not set, the `RATTLER_CACHE_DIR` environment variable is used.
/// - If that is not set, `XDG_CACHE_HOME/pixi` is used when the directory exists.
/// - If that is not set, the default cache directory of [`rattler::default_cache_dir`] is used.
pub fn get_cache_dir() -> miette::Result<PathBuf> {
    std::env::var("PIXI_CACHE_DIR")
        .map(PathBuf::from)
        .or_else(|_| std::env::var("RATTLER_CACHE_DIR").map(PathBuf::from))
        .or_else(|_| {
            let xdg_cache_pixi_dir = std::env::var_os("XDG_CACHE_HOME")
                .map_or_else(
                    || dirs::home_dir().map(|d| d.join(".cache")),
                    |p| Some(PathBuf::from(p)),
                )
                .map(|d| d.join("pixi"));

            // Only use the xdg cache pixi directory when it exists
            xdg_cache_pixi_dir
                .and_then(|d| d.exists().then_some(d))
                .ok_or_else(|| miette::miette!("could not determine xdg cache directory"))
        })
        .or_else(|_| {
            rattler::default_cache_dir()
                .map_err(|_| miette::miette!("could not determine default cache directory"))
        })
}
#[derive(Parser, Debug, Default, Clone)]
pub struct ConfigCli {
    /// Do not verify the TLS certificate of the server.
    #[arg(long, action = ArgAction::SetTrue)]
    tls_no_verify: bool,

    /// Path to the file containing the authentication token.
    #[arg(long, env = "RATTLER_AUTH_FILE")]
    auth_file: Option<PathBuf>,
}

#[derive(Parser, Debug, Default, Clone)]
pub struct ConfigCliPrompt {
    #[clap(flatten)]
    config: ConfigCli,

    /// Do not change the PS1 variable when starting a prompt.
    #[arg(long)]
    change_ps1: Option<bool>,
}

#[derive(Clone, Default, Debug, Deserialize)]
pub struct RepodataConfig {
    /// Disable JLAP compression for repodata.
    pub disable_jlap: Option<bool>,
    /// Disable bzip2 compression for repodata.
    pub disable_bzip2: Option<bool>,
    /// Disable zstd compression for repodata.
    pub disable_zstd: Option<bool>,
}

#[derive(Clone, Default, Debug, Deserialize)]
pub struct Config {
    #[serde(default)]
    pub default_channels: Vec<String>,

    /// If set to true, pixi will set the PS1 environment variable to a custom value.
    #[serde(default)]
    change_ps1: Option<bool>,

    /// Path to the file containing the authentication token.
    #[serde(default)]
    authentication_override_file: Option<PathBuf>,

    /// If set to true, pixi will not verify the TLS certificate of the server.
    #[serde(default)]
    tls_no_verify: Option<bool>,

    #[serde(default)]
    mirrors: HashMap<Url, Vec<Url>>,

    #[serde(skip)]
    pub loaded_from: Vec<PathBuf>,

    #[serde(skip)]
    pub channel_config: ChannelConfig,

    /// Configuration for repodata fetching.
    pub repodata_config: Option<RepodataConfig>,
}

impl From<ConfigCli> for Config {
    fn from(cli: ConfigCli) -> Self {
        Self {
            tls_no_verify: if cli.tls_no_verify { Some(true) } else { None },
            authentication_override_file: cli.auth_file,
            ..Default::default()
        }
    }
}

impl From<ConfigCliPrompt> for Config {
    fn from(cli: ConfigCliPrompt) -> Self {
        let mut config: Config = cli.config.into();
        config.change_ps1 = cli.change_ps1;
        config
    }
}

impl Config {
    /// Parse the given toml string and return a Config instance.
    pub fn from_toml(toml: &str, location: &Path) -> miette::Result<Config> {
        let mut config: Config = toml_edit::de::from_str(toml)
            .into_diagnostic()
            .context(format!("Failed to parse {}", consts::CONFIG_FILE))?;

        config.loaded_from.push(location.to_path_buf());

        Ok(config)
    }

    /// Load the global config file from the home directory (~/.pixi/config.toml)
    pub fn load_global() -> Config {
        let xdg_config_home = std::env::var_os("XDG_CONFIG_HOME").map_or_else(
            || dirs::home_dir().map(|d| d.join(".config")),
            |p| Some(PathBuf::from(p)),
        );

        let global_locations = vec![
            xdg_config_home.map(|d| d.join("pixi").join(consts::CONFIG_FILE)),
            dirs::config_dir().map(|d| d.join("pixi").join(consts::CONFIG_FILE)),
            home_path().map(|d| d.join(consts::CONFIG_FILE)),
        ];
        let mut merged_config = Config::default();
        for location in global_locations.into_iter().flatten() {
            if location.exists() {
                tracing::info!("Loading global config from {}", location.display());
                let global_config = fs::read_to_string(&location).unwrap_or_default();
                if let Ok(config) = Config::from_toml(&global_config, &location) {
                    merged_config = merged_config.merge_config(config);
                } else {
                    tracing::warn!(
                        "Could not load global config (invalid toml): {}",
                        location.display()
                    );
                }
            } else {
                tracing::info!("Global config not found at {}", location.display());
            }
        }

        // Load the default CLI config and layer it on top of the global config
        // This will add any environment variables defined in the `clap` attributes to the config
        let mut default_cli = ConfigCli::default();
        default_cli.update_from(std::env::args().take(0));
        merged_config.merge_config(default_cli.into())
    }

    /// Load the global config and layer the given cli config on top of it.
    pub fn with_cli_config(cli: &ConfigCli) -> Config {
        let config = Config::load_global();
        config.merge_config(cli.clone().into())
    }

    /// Load the config from the given path pixi folder and merge it with the global config.
    pub fn load(p: &Path) -> miette::Result<Config> {
        let local_config = p.join(consts::CONFIG_FILE);
        let mut config = Self::load_global();

        if local_config.exists() {
            let s = fs::read_to_string(&local_config).into_diagnostic()?;
            let local = Config::from_toml(&s, &local_config)?;
            config = config.merge_config(local);
        }

        Ok(config)
    }

    pub fn from_path(p: &Path) -> miette::Result<Config> {
        let s = fs::read_to_string(p).into_diagnostic()?;
        Config::from_toml(&s, p)
    }

    /// Merge the given config into the current one.
    #[must_use]
    pub fn merge_config(mut self, other: Config) -> Self {
        self.mirrors.extend(other.mirrors);
        self.loaded_from.extend(other.loaded_from);

        Self {
            default_channels: if other.default_channels.is_empty() {
                self.default_channels
            } else {
                other.default_channels
            },
            tls_no_verify: other.tls_no_verify.or(self.tls_no_verify),
            change_ps1: other.change_ps1.or(self.change_ps1),
            authentication_override_file: other
                .authentication_override_file
                .or(self.authentication_override_file),
            mirrors: self.mirrors,
            loaded_from: self.loaded_from,
            // currently this is always the default so just use the current value
            channel_config: self.channel_config,
            repodata_config: other.repodata_config.or(self.repodata_config),
        }
    }

    /// Retrieve the value for the default_channels field (defaults to the ["conda-forge"]).
    pub fn default_channels(&self) -> Vec<String> {
        if self.default_channels.is_empty() {
            consts::DEFAULT_CHANNELS
                .iter()
                .map(|s| s.to_string())
                .collect()
        } else {
            self.default_channels.clone()
        }
    }

    /// Retrieve the value for the tls_no_verify field (defaults to false).
    pub fn tls_no_verify(&self) -> bool {
        self.tls_no_verify.unwrap_or(false)
    }

    /// Retrieve the value for the change_ps1 field (defaults to true).
    pub fn change_ps1(&self) -> bool {
        self.change_ps1.unwrap_or(true)
    }

    /// Retrieve the value for the auth_file field.
    pub fn authentication_override_file(&self) -> Option<&PathBuf> {
        self.authentication_override_file.as_ref()
    }

    pub fn channel_config(&self) -> &ChannelConfig {
        &self.channel_config
    }

    pub fn compute_channels(
        &self,
        cli_channels: &[String],
    ) -> Result<Vec<Channel>, ParseChannelError> {
        let channels = if cli_channels.is_empty() {
            self.default_channels()
        } else {
            cli_channels.to_vec()
        };

        channels
            .iter()
            .map(|c| Channel::from_str(c, &self.channel_config))
            .collect::<Result<Vec<Channel>, _>>()
    }

    pub fn mirror_map(&self) -> &std::collections::HashMap<Url, Vec<Url>> {
        &self.mirrors
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_config_parse() {
        let toml = r#"
        default_channels = ["conda-forge"]
        tls_no_verify = true
        "#;
        let config = Config::from_toml(toml, &PathBuf::from("")).unwrap();
        assert_eq!(config.default_channels, vec!["conda-forge"]);
        assert_eq!(config.tls_no_verify, Some(true));
    }

    #[test]
    fn test_config_from_cli() {
        let cli = ConfigCli {
            tls_no_verify: true,
            auth_file: None,
        };
        let config = Config::from(cli);
        assert_eq!(config.tls_no_verify, Some(true));

        let cli = ConfigCli {
            tls_no_verify: false,
            auth_file: Some(PathBuf::from("path.json")),
        };

        let config = Config::from(cli);
        assert_eq!(config.tls_no_verify, None);
        assert_eq!(
            config.authentication_override_file,
            Some(PathBuf::from("path.json"))
        );
    }

    #[test]
    fn test_config_merge() {
        let mut config = Config::default();
        let other = Config {
            default_channels: vec!["conda-forge".to_string()],
            tls_no_verify: Some(true),
            ..Default::default()
        };
        config = config.merge_config(other);
        assert_eq!(config.default_channels, vec!["conda-forge"]);
        assert_eq!(config.tls_no_verify, Some(true));

        let d = Path::new(&env!("CARGO_MANIFEST_DIR"))
            .join("tests")
            .join("config");

        let config_1 = Config::from_path(&d.join("config_1.toml")).unwrap();
        let config_2 = Config::from_path(&d.join("config_2.toml")).unwrap();

        let mut merged = config_1.clone();
        merged = merged.merge_config(config_2);

        let debug = format!("{:#?}", merged);
        let debug = debug.replace("\\\\", "/");
        // replace the path with a placeholder
        let debug = debug.replace(&d.to_str().unwrap().replace('\\', "/"), "path");
        insta::assert_snapshot!(debug);
    }
}
