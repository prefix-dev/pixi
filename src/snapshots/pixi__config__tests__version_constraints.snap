---
source: src/config.rs
expression: results
---
### Strategy: 'Semver'
>=1.2.3,<2 from 1.2.3
>=0.0.0,<0.0.1 from 0.0.0
>=1!1,<1!2 from 1!1
>=1!0.0.0,<1!0.0.1 from 1!0.0.0
>=1.2.3a1,<2 from 1.2.3a1
>=1.2.3a1,<2 from 1.2.3a1, 1.2.3
>=1.2.3a1,<10001 from 1.2.3, 1.2.3a1, 1.2.3b1, 1.2.3rc1, 2, 10000
>=1.2.0,<2 from 1.2.0, 1.3.0
>=0.2.0,<0.4 from 0.2.0, 0.3.0
>=0.2.0,<2 from 0.2.0, 1.3.0
>=1.2,<2 from 1.2
>=1.2,<3 from 1.2, 2
>=1.2,<1!3 from 1.2, 1!2.0

### Strategy: 'Major'
>=1.2.3,<2 from 1.2.3
>=0.0.0,<1 from 0.0.0
>=1!1,<1!2 from 1!1
>=1!0.0.0,<1!1 from 1!0.0.0
>=1.2.3a1,<2 from 1.2.3a1
>=1.2.3a1,<2 from 1.2.3a1, 1.2.3
>=1.2.3a1,<10001 from 1.2.3, 1.2.3a1, 1.2.3b1, 1.2.3rc1, 2, 10000
>=1.2.0,<2 from 1.2.0, 1.3.0
>=0.2.0,<1 from 0.2.0, 0.3.0
>=0.2.0,<2 from 0.2.0, 1.3.0
>=1.2,<2.2 from 1.2
>=1.2,<3 from 1.2, 2
>=1.2,<1!3.0 from 1.2, 1!2.0

### Strategy: 'Minor'
>=1.2.3,<1.3 from 1.2.3
>=0.0.0,<0.1 from 0.0.0
>=1!1,<1!1.1 from 1!1
>=1!0.0.0,<1!0.1 from 1!0.0.0
>=1.2.3a1,<1.3 from 1.2.3a1
>=1.2.3a1,<1.3 from 1.2.3a1, 1.2.3
>=1.2.3a1,<10000.1 from 1.2.3, 1.2.3a1, 1.2.3b1, 1.2.3rc1, 2, 10000
>=1.2.0,<1.4 from 1.2.0, 1.3.0
>=0.2.0,<0.4 from 0.2.0, 0.3.0
>=0.2.0,<1.4 from 0.2.0, 1.3.0
>=1.2,<1.1 from 1.2
>=1.2,<2.1 from 1.2, 2
>=1.2,<1!2.1 from 1.2, 1!2.0

### Strategy: 'ExactVersion'
==1.2.3 from 1.2.3
==0.0.0 from 0.0.0
==1!1 from 1!1
==1!0.0.0 from 1!0.0.0
==1.2.3a1 from 1.2.3a1
==1.2.3a1|==1.2.3 from 1.2.3a1, 1.2.3
==1.2.3|==1.2.3a1|==1.2.3b1|==1.2.3rc1|==2|==10000 from 1.2.3, 1.2.3a1, 1.2.3b1, 1.2.3rc1, 2, 10000
==1.2.0|==1.3.0 from 1.2.0, 1.3.0
==0.2.0|==0.3.0 from 0.2.0, 0.3.0
==0.2.0|==1.3.0 from 0.2.0, 1.3.0
==1.2 from 1.2
==1.2|==2 from 1.2, 2
==1.2|==1!2.0 from 1.2, 1!2.0

### Strategy: 'LatestUp'
>=1.2.3 from 1.2.3
>=0.0.0 from 0.0.0
>=1!1 from 1!1
>=1!0.0.0 from 1!0.0.0
>=1.2.3a1 from 1.2.3a1
>=1.2.3a1 from 1.2.3a1, 1.2.3
>=1.2.3a1 from 1.2.3, 1.2.3a1, 1.2.3b1, 1.2.3rc1, 2, 10000
>=1.2.0 from 1.2.0, 1.3.0
>=0.2.0 from 0.2.0, 0.3.0
>=0.2.0 from 0.2.0, 1.3.0
>=1.2 from 1.2
>=1.2 from 1.2, 2
>=1.2 from 1.2, 1!2.0

### Strategy: 'NoPin'
* from 1.2.3
* from 0.0.0
* from 1!1
* from 1!0.0.0
* from 1.2.3a1
* from 1.2.3a1, 1.2.3
* from 1.2.3, 1.2.3a1, 1.2.3b1, 1.2.3rc1, 2, 10000
* from 1.2.0, 1.3.0
* from 0.2.0, 0.3.0
* from 0.2.0, 1.3.0
* from 1.2
* from 1.2, 2
* from 1.2, 1!2.0
